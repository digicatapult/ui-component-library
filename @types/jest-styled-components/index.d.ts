declare module 'jest-styled-components/serializer'
