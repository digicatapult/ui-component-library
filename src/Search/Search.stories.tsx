import React from 'react'
import { Search } from '../index.js'
import { Story } from '@storybook/react'
import styled from 'styled-components'

export default {
  title: 'Components/Search',
  component: Search,
}

const DefaultStoryTemplate: Story = (args) => {
  return (
    <SearchHost background={args.color} containerWidth={args.containerWidth}>
      <Search
        placeholder={args.placeholder}
        color={args.color}
        background={args.background}
      />
    </SearchHost>
  )
}
export const Default = DefaultStoryTemplate.bind({})
Default.args = {
  placeholder: 'Hello',
  color: '#216968',
  background: 'white',
  containerWidth: 'min-content',
}

interface SearchHostProps {
  background: string
  containerWidth: string
}
const SearchHost = styled.div<SearchHostProps>`
  display: flex;
  justify-content: center;
  width: ${({ containerWidth }) => containerWidth};
  background: ${({ background }) => background};
`
