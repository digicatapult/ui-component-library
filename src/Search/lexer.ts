import moo from 'moo'

import { SearchFields, SearchTerm } from './index.js'

const fieldValidator = /^[a-zA-Z]+$/
export const buildLexer = (fields: SearchFields) => {
  const fieldsArray = Object.entries(fields)
    .map(([fieldName, fieldValue]) => ({
      fieldName,
      ...fieldValue
    }))
    .filter(({ fieldName }) => {
      const isValid = !!fieldName.match(fieldValidator)
      if (!isValid) {
        console.warn(
          `Invalid field ${fieldName} passed to @digicatapult/ui-component-library Search component. Field will be excluded`
        )
      }
      return isValid
    })

  return moo.states({
    main: {
      fieldMatch: {
        match: fieldsArray.map((f) => `${f.fieldName}:`),
        value: (s) => s.slice(0, -1)
      },
      escape: /\\./,
      strstart: { match: '"', push: 'lit' },
      plusMinus: /[+-]/,
      char: /\S/,
      space: { match: /\s+/, lineBreaks: true }
    },
    lit: {
      escape: /\\./,
      strend: { match: /"(?:\s|$)/, pop: 1, lineBreaks: true },
      char: { match: /./, lineBreaks: true }
    }
  })
}

export const parseSearch = (lexer: moo.Lexer, search: string) => {
  const result: SearchTerm[] = []
  let token: moo.Token | undefined
  let type: 'term' | 'fieldMatch' = 'term'
  let field: string = ''
  let value: string = ''
  let modifier: 'positive' | 'negative' | null = null
  let isQuoted: boolean = false

  const maybeAppendToResult = () => {
    if (value) {
      result.push(
        type === 'term'
          ? {
              type,
              value,
              modifier,
              isQuoted
            }
          : {
              type,
              field,
              value,
              modifier,
              isQuoted
            }
      )
      type = 'term'
      field = ''
      value = ''
      modifier = null
      isQuoted = false
    }
  }

  lexer.reset(search)
  while ((token = lexer.next())) {
    switch (token.type) {
      case 'fieldMatch':
        type = token.type
        field = token.value
        value = ''
        break
      case 'plusMinus':
        modifier = token.value === '+' ? 'positive' : 'negative'
        break
      case 'char':
        value = value + token.value
        break
      case 'escape':
        value = value + token.value.slice(1)
        break
      case 'strend':
      case 'space':
        maybeAppendToResult()
        break
      case 'strstart':
        isQuoted = true
        break
    }
  }
  maybeAppendToResult()
  return result
}
