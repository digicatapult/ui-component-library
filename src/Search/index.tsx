import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import styled from 'styled-components'

import { buildLexer, parseSearch } from './lexer.js'

export interface SearchFields {
  [fieldName: string]: {
    // allows us to implement correct input elements for search input helper box
    fieldType: 'text' | 'number' | 'currency' | 'date'
  }
}

interface SearchTermLit {
  type: 'term'
  value: string
  modifier: 'positive' | 'negative' | null
  isQuoted: boolean
}

interface SearchFieldMatch {
  type: 'fieldMatch'
  field: string
  value: string
  modifier: 'positive' | 'negative' | null
  isQuoted: boolean
}

export type SearchTerm = SearchTermLit | SearchFieldMatch
export type SubmitValue = SearchTerm[]

export interface SearchProps {
  fields?: SearchFields
  placeholder?: string
  color?: string
  background?: string
  debounce?: number
  onSubmit: (value: SubmitValue) => void
}

interface SearchWrapperProps extends React.DOMAttributes<HTMLFormElement> {
  color: string
}

interface SearchInputProps extends React.DOMAttributes<HTMLInputElement> {
  color: string
}

interface SearchIconProps extends React.DOMAttributes<HTMLButtonElement> {
  color: string
}

const Search: React.FC<SearchProps> = ({
  fields = {},
  placeholder,
  color = 'black',
  background = 'transparent',
  debounce = 250,
  onSubmit,
}) => {
  const [search, setSearch] = useState<SubmitValue>([])
  const [hasSubmitted, setHasSubmitted] = useState(true)
  const buttonRef = useRef<HTMLButtonElement | null>(null)

  const lexer = useMemo(() => buildLexer(fields), [fields])

  const handleSearchChange: React.ChangeEventHandler<HTMLInputElement> =
    useCallback(
      (event) => {
        const target = event.target as HTMLInputElement
        const terms = parseSearch(lexer, target.value)
        setSearch(terms)
        setHasSubmitted(false)
      },
      [setSearch, setHasSubmitted, lexer],
    )

  const handleSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement> | null) => {
      if (event !== null) {
        event.preventDefault()
        buttonRef.current?.blur()
      }
      if (!hasSubmitted) {
        onSubmit(search)
        setHasSubmitted(true)
      }
    },
    [search, hasSubmitted, setHasSubmitted, onSubmit],
  )

  useEffect(() => {
    const timeoutHandler = setTimeout(() => {
      handleSubmit(null)
    }, debounce)

    return () => clearTimeout(timeoutHandler)
  }, [debounce, search, handleSubmit])

  return (
    <SearchWrapper color={background} onSubmit={handleSubmit}>
      <SearchInput
        placeholder={placeholder}
        type="search"
        name="search"
        color={color}
        onChange={handleSearchChange}
      ></SearchInput>
      <SearchIcon color={color} ref={buttonRef} />
    </SearchWrapper>
  )
}

const SearchWrapper = styled.form<SearchWrapperProps>`
  display: flex;
  width: 100%;
  min-width: 250px;
  align-items: center;
  justify-content: space-between;
  border: 1px solid black;
  border-radius: 10px;
  margin: 1em;
  background: ${({ color }) => color};
  > * {
    margin: 0.5em;
    padding: 0.5em;
  }
`

const SearchInput = styled.input<SearchInputProps>`
  flex-grow: 1;
  flex-shrink: 1;
  border: 0;
  font: inherit;

  &:focus {
    box-shadow: inset 0 0 0 1px ${({ color }) => color};
    border-radius: 7px;
    outline: 0;
  }
`

const SearchIcon = styled.button<SearchIconProps>`
  position: relative;
  font: inherit;
  flex-grow: 0;
  flex-shrink: 0;
  width: 2em;
  height: 2em;
  background: transparent;
  border: 0;

  &, &:before, &:after {
    box-sizing: border-box;
  }

  &:focus {
    &:before, &:after {
      box-shadow: 0 0 5px 1px ${({ color }) => color};
    }
    outline: 0;
  }

  &:before, &:after {
    content: '';
    display: block;
    position: absolute;
  }

  &:before {
    left: 0;
    top: 0;
    width: 80%;
    height: 80%;
    border: calc(4em / 16) solid ${({ color }) => color};
    border-radius: 50%;
  }

  &:after {
    right: 0;
    bottom: 0;
    height: calc(4em / 16);
    width calc(1.41 * 100% - 1.41 * 40% - 40%);
    background: ${({ color }) => color};
    transform: rotateZ(40deg);
    transform-origin: bottom right
  }

  &:hover {
    cursor: pointer;
  }
`

export default Search
