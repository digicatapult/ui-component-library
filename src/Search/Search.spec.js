import React from 'react'
import renderer, { act } from 'react-test-renderer'
import 'jest-styled-components'

import Search from './index.tsx'

describe('Test', () => {
  jest.useFakeTimers()

  test('default', () => {
    const submitEvents = []
    const instance = renderer.create(
      <Search onSubmit={(v) => submitEvents.push(v)} />
    )

    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with props', () => {
    const submitEvents = []
    const instance = renderer.create(
      <Search
        placeholder="placeholder"
        color="green"
        background="red"
        onSubmit={(v) => submitEvents.push(v)}
      />
    )

    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('auto-submit (immediate)', async () => {
    const submitEvents = []
    const instance = renderer.create(
      <Search
        placeholder="placeholder"
        color="green"
        background="red"
        onSubmit={(v) => submitEvents.push(v)}
      />
    )

    await act(() => {
      instance.root
        .findByType('input')
        .props.onChange({ target: { value: 'test' } })
    })

    const tree = instance.toJSON()
    expect(submitEvents).toEqual([])
  })

  test('auto-submit (after 500ms)', async () => {
    const submitEvents = []
    const instance = renderer.create(
      <Search
        placeholder="placeholder"
        color="green"
        background="red"
        onSubmit={(v) => submitEvents.push(v)}
      />
    )

    await act(() => {
      instance.root
        .findByType('input')
        .props.onChange({ target: { value: 'test' } })
    })
    await act(() => {
      jest.advanceTimersByTime(500)
    })

    const tree = instance.toJSON()
    expect(submitEvents).toEqual(['test'])
  })

  test('manual-submit (immediate)', async () => {
    const submitEvents = []
    const instance = renderer.create(
      <Search
        placeholder="placeholder"
        color="green"
        background="red"
        onSubmit={(v) => submitEvents.push(v)}
      />
    )

    await act(() => {
      instance.root
        .findByType('input')
        .props.onChange({ target: { value: 'test' } })
    })
    await act(() => {
      instance.root.findByType('form').props.onSubmit(null)
    })

    const tree = instance.toJSON()
    expect(submitEvents).toEqual(['test'])
  })

  test('manual-submit (wait 500ms)', async () => {
    const submitEvents = []
    const instance = renderer.create(
      <Search
        placeholder="placeholder"
        color="green"
        background="red"
        onSubmit={(v) => submitEvents.push(v)}
      />
    )

    await act(() => {
      instance.root
        .findByType('input')
        .props.onChange({ target: { value: 'test' } })
    })
    await act(() => {
      instance.root.findByType('form').props.onSubmit(null)
    })
    await act(() => {
      jest.advanceTimersByTime(500)
    })

    const tree = instance.toJSON()
    expect(submitEvents).toEqual(['test'])
  })
})
