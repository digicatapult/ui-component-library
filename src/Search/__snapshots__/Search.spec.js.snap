// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  min-width: 250px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border: 1px solid black;
  border-radius: 10px;
  margin: 1em;
  background: transparent;
}

.c0 > * {
  margin: 0.5em;
  padding: 0.5em;
}

.c1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  border: 0;
  font: inherit;
}

.c1:focus {
  box-shadow: inset 0 0 0 1px black;
  border-radius: 7px;
  outline: 0;
}

.c2 {
  position: relative;
  font: inherit;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 2em;
  height: 2em;
  box-sizing: border-box;
  background: transparent;
  border: 0;
}

.c2:focus {
  outline: 0;
}

.c2:focus::before {
  box-shadow: 0 0 5px 1px black;
}

.c2:focus::after {
  box-shadow: 0 0 5px 1px black;
}

.c2::before {
  content: '';
  display: block;
  position: absolute;
  left: 0;
  top: 0;
  width: calc(80% - 2 * 4em / 16);
  height: calc(80% - 2 * 4em / 16);
  border: calc(4em / 16) solid black;
  border-radius: 50%;
}

.c2::after {
  content: '';
  display: block;
  position: absolute;
  right: 0;
  bottom: 0;
  height: calc(4em / 16);
  width: calc(1.41 * 100% - 1.41 * 40% - 40%);
  background: black;
  -webkit-transform: rotateZ(40deg);
  -ms-transform: rotateZ(40deg);
  transform: rotateZ(40deg);
  -webkit-transform-origin: bottom right;
  -ms-transform-origin: bottom right;
  transform-origin: bottom right;
}

.c2:hover {
  cursor: pointer;
}

<form
  className="c0"
  color="transparent"
  onSubmit={[Function]}
>
  <input
    className="c1"
    color="black"
    name="search"
    onChange={[Function]}
    type="search"
  />
  <button
    className="c2"
    color="black"
  />
</form>
`;

exports[`Test with props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  min-width: 250px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border: 1px solid black;
  border-radius: 10px;
  margin: 1em;
  background: red;
}

.c0 > * {
  margin: 0.5em;
  padding: 0.5em;
}

.c1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  border: 0;
  font: inherit;
}

.c1:focus {
  box-shadow: inset 0 0 0 1px green;
  border-radius: 7px;
  outline: 0;
}

.c2 {
  position: relative;
  font: inherit;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 2em;
  height: 2em;
  box-sizing: border-box;
  background: transparent;
  border: 0;
}

.c2:focus {
  outline: 0;
}

.c2:focus::before {
  box-shadow: 0 0 5px 1px green;
}

.c2:focus::after {
  box-shadow: 0 0 5px 1px green;
}

.c2::before {
  content: '';
  display: block;
  position: absolute;
  left: 0;
  top: 0;
  width: calc(80% - 2 * 4em / 16);
  height: calc(80% - 2 * 4em / 16);
  border: calc(4em / 16) solid green;
  border-radius: 50%;
}

.c2::after {
  content: '';
  display: block;
  position: absolute;
  right: 0;
  bottom: 0;
  height: calc(4em / 16);
  width: calc(1.41 * 100% - 1.41 * 40% - 40%);
  background: green;
  -webkit-transform: rotateZ(40deg);
  -ms-transform: rotateZ(40deg);
  transform: rotateZ(40deg);
  -webkit-transform-origin: bottom right;
  -ms-transform-origin: bottom right;
  transform-origin: bottom right;
}

.c2:hover {
  cursor: pointer;
}

<form
  className="c0"
  color="red"
  onSubmit={[Function]}
>
  <input
    className="c1"
    color="green"
    name="search"
    onChange={[Function]}
    placeholder="placeholder"
    type="search"
  />
  <button
    className="c2"
    color="green"
  />
</form>
`;
